master fails to compile with `default-features=false`

**Describe the bug**
Creating a new project with this `Cargo.toml` fails to compile:

```
[package]
name = "foo"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
arrow = {git = "https://github.com/apache/arrow-rs", rev = "80a57be3215ed7210971a501b10664cadfd59bea", default-features=false}
``` 

Output:

```
error[E0432]: unresolved import `rand::prelude::StdRng`
  --> /home/ritchie46/.cargo/git/checkouts/arrow-rs-3b86e19e889d5acc/80a57be/arrow/src/util/bench_util.rs:27:5
   |
27 |     prelude::StdRng,
   |     ^^^^^^^^^^^^^^^ no `StdRng` in `prelude`

error[E0432]: unresolved import `rand::rngs::StdRng`
  --> /home/ritchie46/.cargo/git/checkouts/arrow-rs-3b86e19e889d5acc/80a57be/arrow/src/util/test_util.rs:20:12
   |
20 | use rand::{rngs::StdRng, Rng, SeedableRng};
   |            ^^^^^^^^^^^^ no `StdRng` in `rngs`

error[E0308]: mismatched types
   --> /home/ritchie46/.cargo/git/checkouts/arrow-rs-3b86e19e889d5acc/80a57be/arrow/src/util/bench_util.rs:102:35
    |
94  | pub fn create_string_array<Offset: StringOffsetSizeTrait>(
    |                            ------ this type parameter
...
102 |             if rng.gen::<f32>() < null_density {
    |                                   ^^^^^^^^^^^^ expected type parameter `Offset`, found `f32`
    |
    = note: expected type parameter `Offset`
                         found type `f32`

error[E0308]: mismatched types
   --> /home/ritchie46/.cargo/git/checkouts/arrow-rs-3b86e19e889d5acc/80a57be/arrow/src/util/bench_util.rs:123:35
    |
114 | pub fn create_binary_array<Offset: BinaryOffsetSizeTrait>(
    |                            ------ this type parameter
...
123 |             if rng.gen::<f32>() < null_density {
    |                                   ^^^^^^^^^^^^ expected type parameter `Offset`, found `f32`
    |
    = note: expected type parameter `Offset`
                         found type `f32`

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0432.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `arrow`
```


This is introduced in 80a57be3215ed7210971a501b10664cadfd59bea /  #488

There are quite often feature tag related bugs that could be prevented if we'd run [cargo hack](https://github.com/taiki-e/cargo-hack) in CI. 

Would you be willing to accept a PR that adds those checks? @alamb @jorgecarleitao @Dandandan @jhorstmann @nevi-me

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -244,6 +244,7 @@ jobs:
           rustup component add rustfmt
       - name: Run
         run: cargo fmt --all -- --check
+
   coverage:
     name: Coverage
     runs-on: ubuntu-latest
@@ -327,9 +328,9 @@ jobs:
           cd arrow
           cargo build --target wasm32-unknown-unknown
 
-  # test the projects can build without default features
+  # test builds with various feature flags
   default-build:
-    name: Check No Defaults on AMD64 Rust ${{ matrix.rust }}
+    name: Arrow Feature Flag Builds ${{ matrix.rust }}
     runs-on: ubuntu-latest
     strategy:
       matrix:
@@ -338,15 +339,10 @@ jobs:
     container:
       image: ${{ matrix.arch }}/rust
       env:
-        # Disable full debug symbol generation to speed up CI build and keep memory down
-        # "1" means line tables only, which is useful for panic tracebacks.
-        RUSTFLAGS: "-C debuginfo=1"
-        ARROW_TEST_DATA: /__w/arrow-rs/arrow-rs/testing/data
-        PARQUET_TEST_DATA: /__w/arrow/arrow/parquet-testing/data
+        # Disable debug symbol generation to speed up CI build and keep memory down
+        RUSTFLAGS: "-C debuginfo=0"
     steps:
       - uses: actions/checkout@v2
-        with:
-          submodules: true
       - name: Cache Cargo
         uses: actions/cache@v2
         with:
@@ -357,15 +353,22 @@ jobs:
         uses: actions/cache@v2
         with:
           path: /github/home/target
-          key: ${{ runner.os }}-${{ matrix.arch }}-target-wasm32-cache-${{ matrix.rust }}
+          # this key equals the ones on `linux-build-lib` for re-use
+          key: ${{ runner.os }}-${{ matrix.arch }}-target-cache-${{ matrix.rust }}
       - name: Setup Rust toolchain
         run: |
           rustup toolchain install ${{ matrix.rust }}
           rustup override set ${{ matrix.rust }}
           rustup component add rustfmt
-      - name: Build arrow crate
+      - name: Build with default features
         run: |
           export CARGO_HOME="/github/home/.cargo"
           export CARGO_TARGET_DIR="/github/home/target"
-          cd arrow
-          cargo check --all-targets --no-default-features
+          cd arrow/test/dependency/default-features
+          cargo check
+      - name: Build with default-features=false
+        run: |
+          export CARGO_HOME="/github/home/.cargo"
+          export CARGO_TARGET_DIR="/github/home/target"
+          cd arrow/test/dependency/no-default-features
+          cargo check

--- a/arrow/test/dependency/README.md
+++ b/arrow/test/dependency/README.md
@@ -1,0 +1,21 @@
+<!---
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing,
+  software distributed under the License is distributed on an
+  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  KIND, either express or implied.  See the License for the
+  specific language governing permissions and limitations
+  under the License.
+-->
+
+This directory contains projects that use arrow as a dependency with
+various combinations of feature flags.

--- a/arrow/test/dependency/default-features/Cargo.toml
+++ b/arrow/test/dependency/default-features/Cargo.toml
@@ -1,0 +1,32 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+[package]
+name = "defeault-features"
+description = "Models a user application of arrow that uses default features of arrow"
+version = "0.1.0"
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+arrow = { path = "../../../../arrow", version = "5.0.0-SNAPSHOT" }
+
+# Workaround for https://github.com/apache/arrow-rs/issues/529
+rand = { version = "0.8" }
+
+[workspace]

--- a/arrow/test/dependency/default-features/src/main.rs
+++ b/arrow/test/dependency/default-features/src/main.rs
@@ -1,0 +1,3 @@
+fn main() {
+    println!("Hello, world!");
+}

--- a/arrow/test/dependency/no-default-features/Cargo.toml
+++ b/arrow/test/dependency/no-default-features/Cargo.toml
@@ -1,0 +1,32 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+[package]
+name = "no-default-features"
+description = "Models a user application of arrow that specifies no-default-features=true"
+version = "0.1.0"
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+arrow = { path = "../../../../arrow", version = "5.0.0-SNAPSHOT", default-features = false }
+
+# Workaround for https://github.com/apache/arrow-rs/issues/529
+rand = { version = "0.8" }
+
+[workspace]

--- a/arrow/test/dependency/no-default-features/src/main.rs
+++ b/arrow/test/dependency/no-default-features/src/main.rs
@@ -1,0 +1,3 @@
+fn main() {
+    println!("Hello, world!");
+}

--- a/dev/release/rat_exclude_files.txt
+++ b/dev/release/rat_exclude_files.txt
@@ -3,6 +3,7 @@ testing/*
 target/*
 dev/release/rat_exclude_files.txt
 arrow/test/data/*
+arrow/test/dependency/*
 .gitattributes
 **.gitignore
 .gitmodules

